tools/serverpod_cli/bin/generator/types.dart

```diff
@@ -0,0 +1,348 @@
+import 'package:analyzer/dart/element/nullability_suffix.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:code_builder/code_builder.dart';
+
+import 'class_generator_dart.dart';
+import 'config.dart';
+
+/// Contains information about the type of fields, arguments and return values.
+class TypeDefinition {
+  /// The class name of the type.
+  final String className;
+
+  /// The generics the type has.
+  final List<TypeDefinition> generics;
+
+  final String? url;
+
+  /// Whether this type is nullable.
+  final bool nullable;
+
+  final DartType? dartType;
+
+  /// True if this type references a custom class.
+  final bool customClass;
+
+  bool isEnum;
+
+  TypeDefinition({
+    required this.className,
+    this.generics = const [],
+    required this.nullable,
+    this.url,
+    this.dartType,
+    this.customClass = false,
+    this.isEnum = false,
+  });
+
+  /// Creates an [TypeDefinition] from [mixed] where the [url]
+  /// and [className] is separated by ':'.
+  factory TypeDefinition.mixedUrlAndClassName({
+    required String mixed,
+    List<TypeDefinition> generics = const [],
+    required bool nullable,
+    bool customClass = false,
+  }) {
+    var parts = mixed.split(':');
+    var classname = parts.last;
+    var url = mixed != 'ByteData'
+        ? (parts..removeLast()).join(':')
+        : 'dart:typed_data';
+
+    return TypeDefinition(
+      className: classname,
+      nullable: nullable,
+      generics: generics,
+      url: url.isNotEmpty ? url : null,
+      customClass: customClass,
+    );
+  }
+
+  factory TypeDefinition.fromDartType(DartType type) {
+    var generics = (type is ParameterizedType)
+        ? type.typeArguments.map((e) => TypeDefinition.fromDartType(e)).toList()
+        : <TypeDefinition>[];
+    var url = type.element2?.librarySource?.uri.toString();
+    var nullable = type.nullabilitySuffix == NullabilitySuffix.question;
+
+    var className = !type.isVoid ? type.element2!.displayName : 'void';
+
+    return TypeDefinition(
+      className: className,
+      nullable: nullable,
+      dartType: type,
+      generics: generics,
+      url: url,
+    );
+  }
+
+  static TypeDefinition int = TypeDefinition(className: 'int', nullable: false);
+
+  TypeDefinition get asNullable => TypeDefinition(
+        className: className,
+        url: url,
+        nullable: true,
+        customClass: customClass,
+        dartType: dartType,
+        generics: generics,
+        isEnum: isEnum,
+      );
+
+  /// Generate a [TypeReference] from this definition.
+  TypeReference reference(bool serverCode, [bool? nullable]) => TypeReference(
```

Comment:
Please use () {} syntax when the method is longer than just a single line.

Reply:
7e9f141